ember-client/shared/src/main/scala/org/http4s/ember/client/EmberClientBuilder.scala

```diff
@@ -103,44 +103,88 @@ final class EmberClientBuilder[F[_]: Async] private (
     copy(tlsContextOpt = tlsContext.some)
+
+  /** Unset any `TLSContext` and creates one from the system default `SSLContext`. */
   def withoutTLSContext = copy(tlsContextOpt = None)
 
+  /** Sets the `SocketGroup`, a group of TCP sockets to be used in connections. */
   def withSocketGroup(sg: SocketGroup[F]) = copy(sgOpt = sg.some)
 
+  /** Sets the connection pool's total maximum number of idle connections.
+    * Per `RequestKey` values set with `withMaxPerKey` cannot override this total maximum.
+    */
   def withMaxTotal(maxTotal: Int) = copy(maxTotal = maxTotal)
+
+  /** Sets the connection pool's maximum number of pooled connections per RequestKey. */
   def withMaxPerKey(maxPerKey: RequestKey => Int) = copy(maxPerKey = maxPerKey)
+
+  /** Sets the connection pool's maximum time a connection can be idle. */
   def withIdleTimeInPool(idleTimeInPool: Duration) = copy(idleTimeInPool = idleTimeInPool)
+
+  /** Sets the idle timeout on connections. */
   def withIdleConnectionTime(idleConnectionTime: Duration) =
     copy(idleConnectionTime = idleConnectionTime)
 
+  /** Sets the `Logger`. */
   def withLogger(logger: Logger[F]) = copy(logger = logger)
+
+  /** Sets the max `chunkSize` in bytes to read from sockets at a time. */
   def withChunkSize(chunkSize: Int) = copy(chunkSize = chunkSize)
+
+  /** Sets the max size in bytes the to read from sockets. */
   def withMaxResponseHeaderSize(maxResponseHeaderSize: Int) =
     copy(maxResponseHeaderSize = maxResponseHeaderSize)
 
+  /** Sets the header receive timeout on connections. */
   def withTimeout(timeout: Duration) = copy(timeout = timeout)
+
+  /** Sets additional socket options to apply to the underlying sockets. */
   def withAdditionalSocketOptions(additionalSocketOptions: List[SocketOption]) =
     copy(additionalSocketOptions = additionalSocketOptions)
 
+  /** Sets the User-Agent string.
+    * Manually setting a `User-Agent` header takes priority over this setting.
+    */
   def withUserAgent(userAgent: `User-Agent`) =
     copy(userAgent = userAgent.some)
+
+  /** Sets the User-Agent string to nothing, no User-Agent header would be sent.
+    * Manually setting a `User-Agent` header takes priority over this setting.
+    */
   def withoutUserAgent =
     copy(userAgent = None)
 
+  /** Sets whether or not to force endpoint authentication/verification on the `TLSContext`.
+    * Enabled by default. When enabled the server's identity will be checked against the server's
+    * certificate during SSL/TLS handshaking. This is important to avoid man in the middle attacks
+    * by confirming server identity against their certificate.
+    */
   def withCheckEndpointAuthentication(checkEndpointIdentification: Boolean) =
     copy(checkEndpointIdentification = checkEndpointIdentification)
 
+  /** Disables endpoint authentication/verification. */
   def withoutCheckEndpointAuthentication = copy(checkEndpointIdentification = false)
 
+  /** Sets the `RetryPolicy`. */
```

Comment:
I would have linked to mention the default behaviour here, but the `RetryPolicy` used by default is custom and defined is a private object:
https://github.com/http4s/http4s/blob/9c7c19c264d79c00a0e305555dbdae3219bf5d07/ember-client/shared/src/main/scala/org/http4s/ember/client/internal/ClientHelpers.scala#L252-L272

Perhaps I could simplify things and say something like:
> The default retries on recoverable errors up to two times

Thoughts?

Reply:
This exists because when a tcp disconnect occurs clients can successfully write, but then reading will fail.

So we retryin order to get rid of our dead connection for a hopefully live one.

Reply:
If we're describing it, it's also worth mentioning it only applies to idempotent requests.

Reply:
@ChristopherDavenport The successful write after a tcp disconnect occurs, is that the nature of tcp or something specific to Ember client / http4s?
I'm trying to gauge how important we think this is to mention.

I wonder what other `RetryPolicy`s are sensible. Should they mostly all consider:
- only retrying idempotent requests
- only retrying on `ClosedChannelException`, `Connection reset by peer`, or `Broken pipe`
- retry at least a minimum of two times

If so, perhaps we could introduce a simpler builder option to simply increase the number of retries?
Something like:
```scala
  /** Sets the maximum number of times to retry a `Request` with the default `RetryPolicy`.
   *  The default `RetryPolicy` retries idempotent requests on recoverable errors up to two times.
   *  For more customization see `withRetryPolicy`. 
   */
  def withMaxRetries(maxRetries: Int) =
    copy(maxRetries = maxRetries)
```

---

Independent of the above (which, if wanted, would happen in a separate PR)
How's this for the current `withRetryPolicy`:

> Sets the `RetryPolicy`
> The default retries idempotent requests on recoverable errors up to two times.

Reply:
Blaze added a `withRetries(Int)` recently.  I initially planned to copy what Ember did, but we didn't have a `Timer`, so we couldn't use the `Retry` middleware on 0.22.  (In 0.23, we had an `Async`, and could.)  Ember's version is more flexible, but Blaze's ends up being a lot more concise to change the only value that is likely to change.  I think a convenience setting like that on Ember could be nice.